minimal_data_sharing/minimal_pendulum_controller.cpp

```diff
@@ -0,0 +1,327 @@
+      };
+
+    setpoint_sub_ = this->create_subscription<pendulum_msgs::msg::JointCommand>(
+      "pendulum_setpoint", 10, controller_command_callback);
+
+    // Set PID parameters
+    declare_parameter("p", 1.0);
+    declare_parameter("i", 0.0);
+    declare_parameter("d", 0.0);
+
+    setParameterEventCallback();
+
+    // Create a lambda function that will fire regularly to publish the next results message.
+    auto logger_publish_callback =
+      [this]() {
+        RCLCPP_INFO(
+          this->get_logger(), "Pendulum command position: %lf", get_command());
+        RCLCPP_INFO(
+          this->get_logger(), "Pendulum state position: %lf", get_sensor_position());
+        RCLCPP_INFO(
+          this->get_logger(), "Pendulum setpoint: %lf", get_setpoint());
+        RCLCPP_INFO(this->get_logger(), "----------------------------");
+      };
+
+    // Add a timer to enable regular publication of results messages.
+    logger_publisher_timer_ = this->create_wall_timer(1s, logger_publish_callback);
+  }
+
+  /// Calculate new command based on new sensor state and PID controller properties.
+  // \param[in] msg Received sensor message.
+  void on_sensor_message(pendulum_msgs::msg::JointState::ConstSharedPtr msg)
+  {
+    if (std::isnan(msg->position)) {
+      throw std::runtime_error("Sensor value was NaN in on_sensor_message callback");
+    }
+    // PID controller algorithm
+    double setpoint = get_setpoint();
+    double error = setpoint - msg->position;
+    // Proportional gain is proportional to error
+    double p_gain = pid_.p * error;
+    // Integral gain is proportional to the accumulation of error
+    i_gain_ = pid_.i * (i_gain_ + error * dt_);
+    // Differential gain is proportional to the change in error
+    double d_gain = pid_.d * (error - last_error_) / dt_;
+    last_error_ = error;
+
+    // Calculate the message based on PID gains
+    double new_position;
+    new_position = msg->position + p_gain + i_gain_ + d_gain;
+    // Enforce positional limits
+    if (new_position > M_PI) {
+      new_position = M_PI;
+    } else if (new_position < 0) {
+      new_position = 0;
+    }
+
+    if (std::isnan(new_position)) {
+      throw std::runtime_error("Resulting command was NaN in on_sensor_message callback");
+    }
+
+    set_command(new_position);
+    set_sensor_position(msg->position);
+  }
+
+  /// Callback when a pendulum JointCommand message is received.
+  // \param[in] msg The incoming message containing the position.
+  void on_pendulum_setpoint(pendulum_msgs::msg::JointCommand::ConstSharedPtr msg)
+  {
+    set_setpoint(msg->position);
+    RCLCPP_INFO(this->get_logger(), "Pendulum set to: '%lf'", msg->position);
+  }
+
+  /// Set the properties of the PID controller (gains and desired output).
+  // \param[in] properties Struct representing the desired properties.
+  void set_pid_properties(const PIDProperties & properties)
+  {
+    // TODO(carlos): set data using a lock-free mechanism
```

Comment:
This still needs to be done, right? As it stands this has a data race, no?

Reply:
We discussed in the RTWG meeting and maybe a try-lock will be easier to implement here.